File: Binary Search Concept/iterative/binary_search.py
class Solution:
    def binary_search(self, arr, target):
        s = 0
        e = len(arr) - 1

        while s <= e:
            mid = s + (e - s) // 2

            if target == arr[mid]:
                return mid

            if target > arr[mid]:
                s = mid + 1
            else:
                e = mid - 1

        return -1


if __name__ == "__main__":
    sol = Solution()
    arr = [2, 4, 6, 8, 12, 14, 18, 20]
    result = sol.binary_search(arr, 18)
    print(f"Element found at index: {result}")

File: Binary Search Concept/iterative/binary_search.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>
#pragma GCC optimize("O3,unroll-loops")
using namespace std;

class Solution {
   public:
    int binarySearch(vector<int> &arr, int target) {
        int s = 0;
        int e = arr.size() - 1;
        int mid = s + (e - s) / 2;

        while (s <= e) {  // jab tak "s" kam h "e" se
            mid = s + (e - s) / 2;

            if (target == arr[mid]) {
                return mid;
            }

            if (target > arr[mid]) {  // target is more | arr[mid] is less
                s = mid + 1;
            } else {
                e = mid - 1;
            }
        }

        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> arr = {2, 4, 6, 8, 12, 14, 18, 20};
    int result = sol.binarySearch(arr, 18);
    cout << "Element found at index: " << result << endl;
    return 0;
}
